File: src/main/java/com/github/nsc/de/shake/parser/node/VariableType.java
package com.github.nsc.de.shake.parser.node;

public class VariableType {

    public static VariableType DYNAMIC = new VariableType(Type.DYNAMIC);
    public static VariableType BYTE = new VariableType(Type.BYTE);
    public static VariableType SHORT = new VariableType(Type.SHORT);
    public static VariableType INTEGER = new VariableType(Type.INTEGER);
    public static VariableType LONG = new VariableType(Type.LONG);
    public static VariableType FLOAT = new VariableType(Type.FLOAT);
    public static VariableType DOUBLE = new VariableType(Type.DOUBLE);
    public static VariableType BOOLEAN = new VariableType(Type.BOOLEAN);
    public static VariableType CHAR = new VariableType(Type.CHAR);
    public static VariableType ARRAY = new VariableType(Type.ARRAY);
    public static VariableType OBJECT = new VariableType(Type.OBJECT);
    public static VariableType VOID = new VariableType(Type.VOID);

    private final Type type;
    private final IdentifierNode subtype;

    public VariableType(IdentifierNode subtype) {
        this.type = Type.OBJECT;
        this.subtype = subtype;
    }

    public VariableType(Type type) {
        this.type = type;
        this.subtype = null;
    }

    public Type getType() {
        return type;
    }

    public IdentifierNode getSubtype() {
        return subtype;
    }

    @Override
    public String toString() {
        return "VariableType{" +
                "type=" + type +
                ", subtype='" + subtype + '\'' +
                '}';
    }

    public enum Type {
        DYNAMIC,
        BYTE,
        SHORT,
        INTEGER,
        LONG,
        FLOAT,
        DOUBLE,
        BOOLEAN,
        CHAR,
        ARRAY,
        OBJECT,
        VOID;

        @Override
        public String toString() {
            return this.name().toLowerCase();
        }
    }
}

File: parser/src/main/kotlin/com/github/nsc/de/shake/parser/node/VariableType.kt
package com.github.nsc.de.shake.parser.node

import com.github.nsc.de.shake.util.json.JSON
import java.util.*

@Suppress("unused")
class VariableType {
    val type: Type
    val subtype: IdentifierNode?

    constructor(subtype: IdentifierNode?) {
        type = Type.OBJECT
        this.subtype = subtype
    }

    constructor(type: Type) {
        this.type = type
        subtype = null
    }

    override fun toString(): String = JSON.stringify(this.json)

    val json: Map<String, *>
        get() = toJson()

    fun toJson(): Map<String, *> =
        mapOf(
            "name" to "VariableType",
            "type" to type,
            "subtype" to subtype,
        )

    enum class Type {
        DYNAMIC, BYTE, SHORT, INTEGER, LONG, FLOAT, DOUBLE, BOOLEAN, CHAR, ARRAY, OBJECT, VOID;

        override fun toString(): String {
            return name.lowercase(Locale.getDefault())
        }
    }

    companion object {
        @JvmField
        var DYNAMIC = VariableType(Type.DYNAMIC)
        @JvmField
        var BYTE = VariableType(Type.BYTE)
        @JvmField
        var SHORT = VariableType(Type.SHORT)
        @JvmField
        var INTEGER = VariableType(Type.INTEGER)
        @JvmField
        var LONG = VariableType(Type.LONG)
        @JvmField
        var FLOAT = VariableType(Type.FLOAT)
        @JvmField
        var DOUBLE = VariableType(Type.DOUBLE)
        @JvmField
        var BOOLEAN = VariableType(Type.BOOLEAN)
        @JvmField
        var CHAR = VariableType(Type.CHAR)
        @JvmField
        var ARRAY = VariableType(Type.ARRAY)
        @JvmField
        var OBJECT = VariableType(Type.OBJECT)
        @JvmField
        var VOID = VariableType(Type.VOID)
    }
}